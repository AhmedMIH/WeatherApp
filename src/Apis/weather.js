import { useQuery } from "react-query";
import Config from "react-native-config";
import { axiosRequest } from "../Services/api.service";


export const fetchWeather = ( location ) => {

    return axiosRequest( { url: 'forecast.json', params: { key: Config.API_KEY, q: location, days: 4, hour: 14 }, method: 'GET' } );
};
export const useGetWeather = ( query ) => {
    return useQuery( [ 'weather', query ], () => { return fetchWeather( query ) }, { cacheTime: 0, retry: 0 } );
}