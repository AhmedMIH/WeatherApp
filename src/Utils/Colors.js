export default {
    white: '#FFFFFF',
    black: '#000000',
    red: '#FF0000',
    green: '#00FF00',
    blue: '#0000FF',
    yellow: '#FFFF00',
    orange: '#FFA500',
    primary: '#32255A',
    border: '#502d6b',
    container: '#F1F2F3',

}