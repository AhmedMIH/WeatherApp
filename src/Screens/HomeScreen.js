import React, { useState } from 'react'
import { Text, StyleSheet } from 'react-native'
import LinearGradient from 'react-native-linear-gradient'
import Spinner from 'react-native-loading-spinner-overlay'
import SearchComponent from '../Components/SearchComponent'
import WeathersList from '../Components/WeathersList'
import { useGetWeather } from '../Apis/weather'
import { responsiveHeight, responsiveWidth } from '../Utils/Helper'
import Colors from '../Utils/Colors'


const HomeScreen = () => {
    const [ searchQuery, setSearchQuery ] = useState( 'London' )
    const { data, isError, isLoading } = useGetWeather( searchQuery )


    const transformWeatherData = ( data ) => {
        if ( !data || !data.forecast || !data.forecast.forecastday ) {
            return [];
        }

        return data.forecast.forecastday.map( item => {
            return {
                date: item.date,
                temp: {
                    maxtemp_c: item.day.maxtemp_c,
                    mintemp_c: item.day.mintemp_c,
                    avgtemp_c: item.day.avgtemp_c,
                    avgvis_km: item.day.avgvis_km,
                    avghumidity: item.day.avghumidity,
                },
                icon: { text: item.day.condition.text, source: item.day.condition.icon },
                astro: {
                    sunrise: item.astro.sunrise,
                    sunset: item.astro.sunset,
                    moonrise: item.astro.moonrise,
                    moonset: item.astro.moonset
                }
            };
        } );
    };


    return (
        <LinearGradient colors={[ Colors.gradient, Colors.gradient2 ]} style={styles.container}>
            <Spinner visible={isLoading} />
            <SearchComponent searchQuery={searchQuery} onChange={setSearchQuery} />
            {isError ? <Text style={styles.errorText}>Something went wrong Please Enter Valid City Or Zip Code</Text> :
                <>
                    <Text style={styles.cityName}>{data?.location?.name}</Text>
                    <WeathersList data={transformWeatherData( data )} />
                </>
            }
        </LinearGradient>
    )
}


const styles = StyleSheet.create( {
    container: {
        flex: 1,
        paddingVertical: responsiveHeight( 16 ),
        paddingHorizontal: responsiveWidth( 16 )
    },
    cityName: {
        fontSize: 24,
        fontWeight: 'bold',
        marginTop: 16,
        color: Colors.white
    },
    errorText: {
        flex: 1,
        fontSize: 24,
        fontWeight: 'bold',
        marginTop: 16,
        color: Colors.white,
        textAlign: 'center',
        textAlignVertical: 'center',
    },
} )
export default HomeScreen