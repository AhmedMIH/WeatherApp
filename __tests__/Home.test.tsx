// HomeScreen.test.js
import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import HomeScreen from '../src/Screens/HomeScreen';
 import Colors from '../src/Utils/Colors';
import { QueryClient, QueryClientProvider } from 'react-query';
import { useGetWeather } from '../src/Apis/weather';
import { NavigationContainer } from '@react-navigation/native';

jest.mock('../src/Apis/weather');


const queryClient = new QueryClient();
const mockWeatherData = {
  location: {
    name: 'Cairo'
  },
  forecast: {
    forecastday: [
      {
        date: '2024-05-26',
        day: {
          maxtemp_c: 35.8,
          mintemp_c: 21.6,
          avgtemp_c: 29.3,
          avgvis_km: 10,
          avghumidity: 26,
          condition: {
            text: 'Sunny',
            icon: '//cdn.weatherapi.com/weather/64x64/day/113.png'
          }
        },
        astro: {
          sunrise: '04:56 AM',
          sunset: '06:48 PM',
          moonrise: '09:58 PM',
          moonset: '06:58 AM'
        }
      }
    ]
  }
};


describe('HomeScreen', () => {
  const renderComponent = () =>
    render(
      <QueryClientProvider client={queryClient}>
        <NavigationContainer>
            <HomeScreen />
        </NavigationContainer>
      </QueryClientProvider>
    );

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', async () => {
    useGetWeather.mockReturnValue({
      data: mockWeatherData,
      isError: false,
      isLoading: false,
    });

    const { getByText } = renderComponent();
    await waitFor(() => expect(getByText('Cairo')).toBeTruthy());
  });

  it('shows loading spinner while fetching data', () => {
    useGetWeather.mockReturnValue({
      data: null,
      isError: false,
      isLoading: true,
    });

    const { getByTestId } = renderComponent();
    expect(getByTestId('loading-spinner')).toBeTruthy();
  });

  it('shows error message on error', async () => {
    useGetWeather.mockReturnValue({
      data: null,
      isError: true,
      isLoading: false,
    });

    const { getByText } = renderComponent();
    await waitFor(() => expect(getByText('Something went wrong Please Enter Valid City Or Zip Code')).toBeTruthy());
  });

  it('updates search query', () => {
    useGetWeather.mockReturnValue({
      data: mockWeatherData,
      isError: false,
      isLoading: false,
    });

    const { getByPlaceholderText } = renderComponent();
    const searchInput = getByPlaceholderText('Enter a city or zip code');
    fireEvent.changeText(searchInput, 'New York');
    expect(searchInput.props.value).toBe('New York');
  });

  it('renders weather data correctly', async () => {
    useGetWeather.mockReturnValue({
      data: mockWeatherData,
      isError: false,
      isLoading: false,
    });

    const { getByText } = renderComponent();
    await waitFor(() => expect(getByText('Sunny')).toBeTruthy());
    await waitFor(() => expect(getByText('Cairo')).toBeTruthy());
    await waitFor(() => expect(getByText('2024-05-26')).toBeTruthy());
    await waitFor(() => expect(getByText('29.3Â°')).toBeTruthy());
  });
});