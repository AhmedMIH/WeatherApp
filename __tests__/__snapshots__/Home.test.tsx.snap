// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders correctly 1`] = `
<View
  colors={
    [
      "#2E335A",
      "#1C1B33",
    ]
  }
  style={
    {
      "flex": 1,
      "paddingHorizontal": 30.76923076923077,
      "paddingVertical": 25.28909952606635,
    }
  }
>
  <View
    testID="loading-spinner"
  />
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 8,
        "height": 75.86729857819905,
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#1C1B33",
          "borderRadius": 8,
          "height": "100%",
          "justifyContent": "center",
          "opacity": 1,
          "paddingHorizontal": 15.384615384615385,
          "paddingVertical": 18.966824644549764,
        }
      }
    >
      <Image
        source={
          {
            "testUri": "../../../src/assets/icons/location.png",
          }
        }
        style={
          {
            "height": 24,
            "width": 24,
          }
        }
      />
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#1C1B33",
          "borderRadius": 8,
          "flex": 1,
          "flexDirection": "row",
          "paddingHorizontal": 23.076923076923077,
        }
      }
    >
      <Image
        source={
          {
            "testUri": "../../../src/assets/icons/search.png",
          }
        }
        style={
          {
            "height": 24,
            "width": 24,
          }
        }
      />
      <TextInput
        onChangeText={[Function]}
        placeholder="Enter a city or zip code"
        placeholderTextColor="#FFFFFF"
        style={
          {
            "color": "#FFFFFF",
            "flex": 1,
            "fontSize": 7,
            "marginHorizontal": 15.384615384615385,
          }
        }
        value=""
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#4B32A1",
            "borderRadius": 8,
            "justifyContent": "center",
            "opacity": 1,
            "paddingHorizontal": 15.384615384615385,
            "paddingVertical": 12.644549763033176,
          }
        }
      >
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 7,
              "fontWeight": "medium",
            }
          }
        >
          Search
        </Text>
      </View>
    </View>
  </View>
  <Text
    style={
      {
        "color": "#FFFFFF",
        "fontSize": 24,
        "fontWeight": "bold",
        "marginTop": 16,
      }
    }
  />
  <RCTScrollView
    contentContainerStyle={
      {
        "gap": 16,
      }
    }
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    style={
      {
        "marginTop": 16,
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
</View>
`;
